### PR TITLE
Pass CancellationToken in compilers layer

### PR DESCRIPTION
**Pull request by jmarolf**
Redo of https://github.com/dotnet/roslyn/pull/52247

This cleans up passing cancellation tokens in the compiler layer

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

**Comment by RikkiGibson** [comment]
I was caught off guard by the merge with only one approval. It looks like this other approval is for the same set of changes as this PR: https://github.com/dotnet/roslyn/pull/52247#pullrequestreview-624527280. So I suppose it is alright.

**Comment by 333fred** [comment]
@jmarolf please remember that compiler PRs require 2 sign offs, not just one.

@dotnet/roslyn-compiler, please take a second, retroactive review. I don't believe that this will need to have any changes, but it does need a second review.

**Pull request merged**
